zuul-core/src/main/java/com/netflix/zuul/netty/timeouts/HttpHeadersTimeoutHandler.java

```diff
@@ -0,0 +1,98 @@
+
+public class HttpHeadersTimeoutHandler {
+        private static final Logger LOG = LoggerFactory.getLogger(HttpHeadersTimeoutHandler.class);
+
+        private static final AttributeKey<ScheduledFuture<Void>> HTTP_HEADERS_READ_TIMEOUT_FUTURE =
+            AttributeKey.newInstance("httpHeadersReadTimeoutFuture");
+
+        public static class InboundHandler extends ChannelInboundHandlerAdapter {
+            private final BooleanSupplier httpHeadersReadTimeoutEnabledSupplier;
+            private final IntSupplier httpHeadersReadTimeoutSupplier;
+
+            private final Counter httpHeadersReadTimeoutCounter;
+
+            private boolean closed = false;
+
+            public InboundHandler(BooleanSupplier httpHeadersReadTimeoutEnabledSupplier, IntSupplier httpHeadersReadTimeoutSupplier, Counter httpHeadersReadTimeoutCounter) {
+                this.httpHeadersReadTimeoutEnabledSupplier = httpHeadersReadTimeoutEnabledSupplier;
+                this.httpHeadersReadTimeoutSupplier = httpHeadersReadTimeoutSupplier;
+                this.httpHeadersReadTimeoutCounter = httpHeadersReadTimeoutCounter;
+            }
+
+            @Override
+            public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                try {
+                    if (!httpHeadersReadTimeoutEnabledSupplier.getAsBoolean())
+                        return;
+                    int timeout = httpHeadersReadTimeoutSupplier.getAsInt();
+                    ctx.channel().attr(HTTP_HEADERS_READ_TIMEOUT_FUTURE).set(
+                        ctx.executor().schedule(
+                            () -> {
+                                if (!closed) {
+                                    ctx.fireExceptionCaught(ReadTimeoutException.INSTANCE);
+                                    ctx.close();
+                                    closed = true;
+                                    httpHeadersReadTimeoutCounter.increment();
+                                    LOG.debug("[{}] HTTP headers read timeout handler timed out", ctx.channel().id());
+                                }
+                                return null;
+                            },
+                            timeout,
+                            TimeUnit.MILLISECONDS
+                        )
+                    );
+                    LOG.debug("[{}] Adding HTTP headers read timeout handler: {}", ctx.channel().id(), timeout);
+                } finally {
+                    super.channelActive(ctx);
+                }
+            }
+
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                try {
+                    if (msg instanceof HttpMessage) {
+                        ScheduledFuture<Void> future = ctx.channel().attr(HTTP_HEADERS_READ_TIMEOUT_FUTURE).get();
+                        if (future != null) {
+                            future.cancel(false);
```

Comment:
in addition to canceling it would be nice for this handler to remove itself from the pipeline once it's no longer doing anything